blinemedical/amazon-s3-gst-plugin
# [VIDEO-1728] Seekable s3sink

*Issue #, if available:* VIDEO-1728

*Description of changes:*
* Moves API handle to it's own header for re-use
* Adds `aws-s3` logging category for API logging
* Adds support for seeking in S3 sink, optimized with copy-uploading

To test run:
```
GST_DEBUG=*:3,s3sink:7 gst-launch-1.0 videotestsrc pattern=snow num-buffers=2100 ! video/x-raw,width=1280,height=720,framerate=30/1 ! x264enc ! video/x-h264,profile=constrained-baseline ! mp4mux ! s3sink bucket=dev-laerdallabs-vera-recordings key=test.mp4
```
You'll see something like this in the logs:
```
0:00:06.829771200   519 0x555c890e4980 DEBUG                 s3sink gsts3sink.c:688:gst_s3_sink_flush_buffer:<s3sink0> Uploading 5242880 byte part
0:00:15.227416800   519 0x555c890e4980 DEBUG                 s3sink gsts3sink.c:688:gst_s3_sink_flush_buffer:<s3sink0> Uploading 5242880 byte part
0:00:23.474965300   519 0x555c890e4980 DEBUG                 s3sink gsts3sink.c:688:gst_s3_sink_flush_buffer:<s3sink0> Uploading 5242880 byte part
0:00:24.407088000   519 0x555c890e4980 DEBUG                 s3sink gsts3sink.c:820:gst_s3_sink_do_seek:<s3sink0> Seeking to new offset 32 from 16563157 of 16563157 total bytes
0:00:24.407151100   519 0x555c890e4980 DEBUG                 s3sink gsts3sink.c:743:gst_s3_sink_do_flush:<s3sink0> Flushing S3 Upload
0:00:24.407238600   519 0x555c890e4980 DEBUG                 s3sink gsts3sink.c:688:gst_s3_sink_flush_buffer:<s3sink0> Uploading 834517 byte part
...
0:00:24.956218200   519 0x555c890e4980 TRACE                 s3sink gsts3sink.c:862:gst_s3_sink_do_seek:<s3sink0> Seeking to offset 32 by downloading
0:00:25.046026200   519 0x555c890e4980 DEBUG                 s3sink gsts3sink.c:743:gst_s3_sink_do_flush:<s3sink0> Flushing S3 Upload
0:00:25.046117300   519 0x555c890e4980 TRACE                 s3sink gsts3sink.c:754:gst_s3_sink_do_flush:<s3sink0> Post-filling 5242832 bytes, range: 48-5242880
0:00:25.235402600   519 0x555c890e4980 DEBUG                 s3sink gsts3sink.c:688:gst_s3_sink_flush_buffer:<s3sink0> Uploading 5242880 byte part
0:00:25.236415100   519 0x555c890e4980 TRACE                 s3sink gsts3sink.c:794:gst_s3_sink_do_flush:<s3sink0> Copy-uploading remaining file from 5242880 to 16563157
```
Here it's uploading 3 5MB chunks, gets a seek to new offset (`32`) which causes it to upload the current/last `834517` bytes then complete the upload. Then it does the actual seek by downloading the first `32` bytes, receives more data (not logged), "post-fills" the rest of the data in the 5MB chunk, uploads the updated 5MB chunk, then copy-uploads (which is performed _in_ S3) the remaining data, and completes the upload.

The resulting file in S3 should be playable.
